File: src_local/utils/context_engine.py
import logging
import re
from pathlib import Path
from src_local.agents.Support.dna_profiler import DNAProfiler
from src_local.agents.Support.coverage_auditor import CoverageAuditor

logger = logging.getLogger(__name__)

class ContextEngine:
    """
    🧠 Cérebro Semântico PhD.
    O Orquestrador de metadados responsável por mapear a topologia do projeto.
    """
    
    def __init__(self, project_root, support_tools=None):
        self.project_root = Path(project_root)
        self.map, self.call_graph, self.project_identity = {}, {}, {}
        self.dna_profiler = DNAProfiler()
        self.coverage_auditor = CoverageAuditor()
        self.all_files_index = []
        
        if support_tools:
            self._inject_support(support_tools)
        else:
            self._initialize_support_tools()

    def _inject_support(self, tools):
        self.analyst, self.guardian = tools["analyst"], tools["guardian"]
        self.mapper, self.parity_analyst = tools["mapper"], tools["parity"]

    def _initialize_support_tools(self):
        from src_local.agents.Support.infrastructure_assembler import InfrastructureAssembler
        support = InfrastructureAssembler.assemble_core_support()
        self._inject_support(support)

    def analyze_project(self):
        """🔭 Varre o projeto delegando a inteligência para os assistentes técnicos."""
        from src_local.utils.file_system_scanner import FileSystemScanner
        scanner = FileSystemScanner(self.project_root, self.analyst)
        
        logger.info("🧠 Mapeando topologia...")
        self.project_identity = self._discover_identity()
        self.map, self.all_files_index = {}, scanner.scan_all_filenames()

        # Processamento em Lote
        analyzable = scanner.get_analyzable_files()
        self._content_cache = self._pre_read_files(analyzable)
        self._register_batch_files()
        
        # Finalização
        self._content_cache = {}
        self._build_dependency_map()
        logger.info(f"✅ DNA Processado: {len(self.map)} componentes.")
        return {"identity": self.project_identity, "map": self.map}

    def _register_batch_files(self):
        """Itera sobre o cache de conteúdo para registrar cada arquivo."""
        for path_str in self._content_cache.keys():
            self._register_file(Path(self.project_root / path_str))

    def _pre_read_files(self, files_iterator):
        cache = {}
        for path in files_iterator:
            try:
                rel = path.relative_to(self.project_root).as_posix()
                cache[rel] = path.read_text(encoding='utf-8', errors='ignore')
            except: continue
        return cache

    def _register_file(self, path, ignore_test_context=False):
        try:
            rel_path = path.relative_to(self.project_root).as_posix()
            if rel_path in self.map: return
            
            content = self._get_cached_content(path, rel_path)
            info = self._get_initial_info(path, rel_path)
            info["content"] = content
            
            self._perform_deep_analysis(path, content, info, ignore_test_context)
            self.map[rel_path] = info
        except Exception as e:
            logger.error(f"❌ Erro ao analisar {path}: {e}")

    def _get_cached_content(self, path, rel_path):
        if not hasattr(self, '_content_cache'): self._content_cache = {}
        content = self._content_cache.get(rel_path)
        return content if content is not None else path.read_text(encoding='utf-8', errors='ignore')

    def _perform_deep_analysis(self, path, content, info, ignore_test):
        if path.suffix == '.py':
            info.update(self.analyst.analyze_python(content, path.name))
        
        # Auditorias de vulnerabilidade e teste
        info.update(self.guardian.detect_vulnerabilities(content, info["component_type"], ignore_test))
        info["has_test"] = self.coverage_auditor.detect_test(path, info["component_type"], self.all_files_index)
        
        if info["component_type"] == "TEST":
            self._analyze_test_quality(content, info)

    def _get_initial_info(self, path: Path, rel_path):
        comp_type = self.analyst.map_component_type(rel_path)
        return {
            "purpose": "Logic", "functions": [], "classes": [],
            "brittle": False, "silent_error": False, "has_test": False,
            "component_type": comp_type, 
            "domain": "EXPERIMENTATION" if comp_type == "TEST" else "PRODUCTION", 
            "path": str(path), "rel_path": rel_path
        }

    def _analyze_test_quality(self, content, info):
        # Mapeia densidade de asserções
        assertions = len(re.findall(r"assert[A-Z]\w*\(|self\.assert|check\(|assertThat\(|expect\(", content))
        assertions += content.count("assert ")
        info["test_depth"] = {"assertion_count": assertions, "quality_level": "DEEP" if assertions > 5 else "SHALLOW"}

    def _build_dependency_map(self):
        for file, data in self.map.items():
            data["coupling"] = self.mapper.calculate_metrics(file, data, self.map)
            self.call_graph[file] = self._find_dependents(file, data)

    def _find_dependents(self, file, data):
        """Localiza componentes que dependem do arquivo atual."""
        return [f for f in self.map.keys() if f != file and any(c in str(data) for c in self.map[f].get('classes', []))]

    def analyze_stack_parity(self, personas):
        parity = self.parity_analyst.analyze_stack_gaps(personas)
        parity["detected"] = self.project_identity.get("stacks", set())
        return parity

    def _discover_identity(self):
        try: return self.dna_profiler.discover_identity(self.project_root)
        except Exception as e:
            logger.error(f"🚨 Falha ao descobrir identidade: {e}")
            return {"stacks": set(), "core_mission": "Software Proposital"}

    def get_criticality_score(self, file_path):
        score = 0
        stem = Path(file_path).stem
        for deps in self.call_graph.values():
            if any(stem in str(d) for d in deps): score += 1
        if any(kw in str(file_path) for kw in ["core", "base"]): score += 10
        return score

File: src_local/utils/context_engine.ts

import { HubManagerGRPC } from "../core/hub_manager_grpc.ts";
import { Path } from "../core/path_utils.ts";
import { DNAProfiler } from "../agents/Support/Analysis/dna_profiler.ts";
import { ContextMappingLogic } from "./context_mapping_logic.ts";
import { FileSystemScanner } from "./file_system_scanner.ts";
import { StructuralAnalyst } from "../agents/Support/Analysis/structural_analyst.ts";
import { CoverageAuditor } from "../agents/Support/Analysis/coverage_auditor.ts";
import { ConnectivityMapper } from "../agents/Support/Analysis/connectivity_mapper.ts";
import { ParityAnalyst } from "../agents/Support/Analysis/parity_analyst.ts";
import { MetricsEngine } from "../agents/Support/Diagnostics/metrics_engine.ts";
import { ContextHelpers } from "./context_helpers.ts";
import type { CognitiveStatus, IAgent } from "../core/types.ts";

/**
 * 🧠 Cérebro Semântico PhD (Bun Version).
 */
export class ContextEngine {
    projectRoot: Path; 
    map: Record<string, any> = {}; 
    callGraph: Record<string, string[]> = {};
    dnaProfiler: DNAProfiler; 
    mappingLogic: ContextMappingLogic;
    analyst: StructuralAnalyst; 
    coverageAuditor = new CoverageAuditor();
    connectivityMapper: ConnectivityMapper; 
    parityAnalyst = new ParityAnalyst();
    metricsEngine = new MetricsEngine(); 
    allFilesIndex: string[] = [];
    projectIdentity: Record<string, unknown> = {};
    private contentCache: Record<string, string> = {};

    constructor(projectRoot: string, private hubManager?: HubManagerGRPC) {
        this.projectRoot = new Path(projectRoot);
        this.dnaProfiler = new DNAProfiler(hubManager);
        this.connectivityMapper = new ConnectivityMapper(hubManager);
        this.analyst = new StructuralAnalyst(hubManager);
        this.mappingLogic = new ContextMappingLogic(hubManager);
    }

    async analyzeProject(): Promise<{ identity: Record<string, unknown>; map: Record<string, any> }> {
        this.projectIdentity = await this.dnaProfiler.discoverIdentity(this.projectRoot);
        const scanner = new FileSystemScanner(this.projectRoot.toString(), this.analyst);
        this.allFilesIndex = await scanner.scanAllFilenames();

        const goMap = await this.getGoDiscoveryMap();
        this.contentCache = await this.mappingLogic.processBatch(scanner, this, goMap);

        await this.buildDependencyMap();
        return { identity: this.projectIdentity, map: this.map };
    }

    private async getGoDiscoveryMap(): Promise<Record<string, any>> {
        const { GoDiscoveryAdapter } = await import("./go_discovery_adapter.ts");
        const { results: goResults } = await GoDiscoveryAdapter.scan(this.projectRoot.toString(), this.projectRoot.toString(), this.hubManager);
        const goMap: Record<string, any> = {};
        goResults.forEach(r => goMap[r.path.replace(/\\/g, "/")] = r);
        return goMap;
    }

    async registerFile(path: Path, ignoreTest: boolean = false, goMetrics?: any) {
        const rel = path.relativeTo(this.projectRoot);
        if (this.map[rel]) return;

        const content = await this.getCachedContent(path, rel);
        const info = this.mappingLogic.getInitialInfo(path, rel, this.analyst);

        const rustMeta = this.mappingLogic.metadataCache[rel.replace(/\\/g, "/")];
        if (rustMeta) {
            info.rust_metadata = rustMeta;
        }

        info.content = content;

        if (goMetrics) {
            info.atomic_go_metrics = {
                totalComplexity: goMetrics.total_complexity,
                cognitiveComplexity: goMetrics.cognitive_complexity,
                maxNesting: goMetrics.max_nesting,
                loc: goMetrics.loc,
                sloc: goMetrics.sloc,
                comments: goMetrics.comments
            };
        }

        await this.performDeepAnalysis(path, content, info, ignoreTest);
        this.map[rel] = info;
    }

    private async getCachedContent(path: Path, rel: string): Promise<string> {
        if (this.contentCache[rel]) return this.contentCache[rel]!;
        return this.readFileContent(path);
    }

    private async readFileContent(path: Path): Promise<string> {
        try {
            return await (Bun as any).file(path.toString()).text();
        } catch {
            return "";
        }
    }

    private async performDeepAnalysis(path: Path, content: string, info: any, ignoreTest: boolean) {
        await this._applyStructuralAnalysis(path, content, info);
        this._applyAdvancedMetrics(path, content, info, info.atomic_go_metrics);
        await this._applySecurityAndTests(path, content, info, ignoreTest);
    }

    private async _applyStructuralAnalysis(path: Path, content: string, info: any) {
        const name = path.name();
        const structural = path.toString().endsWith('.py')
            ? await this.analyst.analyzePython(content, name)
            : await this.analyst.analyze_file_logic(content, name);

        // Add intent classification using Rust metadata if available
        info.intent = this.analyst.analyze_intent(content, name, info.rust_metadata);

        Object.assign(info, structural);
    }

    private _applyAdvancedMetrics(path: Path, content: string, info: any, goMetrics?: any) {
        const rel = path.relativeTo(this.projectRoot).replace(/\\/g, "/");
        const adv = this.metricsEngine.analyzeFile(content, rel, info.dependencies || [], 0, goMetrics);

        Object.assign(info, {
            advanced_metrics: { ...adv },
            telemetry: adv.telemetry || info.telemetry,
            complexity: adv.cyclomaticComplexity
        });

        this.enrichShadowCompliance(info, adv);
    }

    private enrichShadowCompliance(info: any, adv: any) {
        if (adv.isShadow) {
            const v = this.metricsEngine.validateShadowCompliance(adv);
            Object.assign(info, {
                shadow_compliance: { compliant: v.compliant, reason: v.reason },
                complexity: adv.shadowComplexity
            });
        }
    }

    private async _applySecurityAndTests(path: Path, content: string, info: any, ignoreTest: boolean) {
        const vuln = await this.analyst.integrityGuardian.detectVulnerabilities(content, info.component_type, path.name(), ignoreTest);
        Object.assign(info, vuln);

        info.has_test = this.coverageAuditor.detectTest(path, info.component_type, this.allFilesIndex, info);
        this.enrichTestDepth(content, info);
    }

    private enrichTestDepth(content: string, info: any) {
        if (info.component_type === "TEST") {
            const matches = (content.match(/assert|expect|should/g) || []).length;
            info.test_depth = { assertion_count: matches, quality_level: matches > 5 ? "DEEP" : "SHALLOW" };
        }
    }

    private async buildDependencyMap() {
        const bulkResults = await this.connectivityMapper.calculateBulk(this.map);
        const hasBulk = Object.keys(bulkResults).length > 0;

        Object.keys(this.map).forEach(f => {
            if (hasBulk && bulkResults[f]) {
                this.map[f].coupling = bulkResults[f];
            } else {
                this.map[f].coupling = this.connectivityMapper.calculateMetrics(f, this.map[f], this.map);
            }
        });
        this.callGraph = {};
        this.populateCallGraph();
    }

    private populateCallGraph() {
        Object.entries(this.map).forEach(([f, d]) => {
            (d.dependencies || []).forEach((dep: string) => {
                const res = ContextHelpers.resolveDependency(dep, this.map);
                if (res && res !== f) {
                    (this.callGraph[res] ||= []).push(f);
                }
            });
        });
    }

    analyzeStackParity(personas: IAgent[]) { 
        const p = this.parityAnalyst.analyzeStackGaps(personas) as any; 
        p.detected = (this.projectIdentity.stacks as Set<string>) || new Set<string>(); 
        return p; 
    }
    async cognitiveReason(p: string): Promise<CognitiveStatus> { 
        const res = await (await new (await import("./cognitive_engine.ts")).CognitiveEngine()).reason(p); 
        if (typeof res === 'string') return { status: res, score: 0 };
        return res || { status: "Unknown", score: 0 };
    }
    _getScanner() { return new FileSystemScanner(this.projectRoot.toString(), this.analyst); }
    get_criticality_score(f: string) { return ContextHelpers.getCriticalityScore(f, this.map); }
}
